File: server/random.js
/* */

import express from "express";
import { aux } from "./middlewares.js";

import { createRequire } from "module";
const require = createRequire(import.meta.url);
const surnames = require("./data/surnames.json");

const randomRouter = express.Router();
randomRouter.get("/v1/random/name", function (req, res) {
  const response = random.name();
  //console.log(response);
  res.status(200).json(response);
});
randomRouter.get("/v1/random/integer", function (req, res, next) {
  const response = random.integer(req);
  if (!response) {
    const error = new Error(aux.badRequest);
    error.code = 400;
    next(error);
    return;
  }
  res.status(200).json(response);
});
randomRouter.get("/v1/random/list", function (req, res, next) {
  const response = random.list(req);
  if (!response) {
    const error = new Error(aux.badRequest);
    error.code = 400;
    next(error);
    return;
  }
  res.status(200).json(response);
});

export {
  randomRouter
};

const random = {
  response: {
    quest: "",
    data: [],
  },
  clean: function () {
    this.response.quest = "";
    this.response.data = [];
  },
  name: function name() {
    this.clean();
    this.response.quest = "Random Name";
    this.response.data.push(surnames[aux.randomInt(0, surnames.length - 1)]);
    return this.response;
  },
  integer: function (req, res, next) {
    const min = parseInt(req.query.min);
    const max = parseInt(req.query.max);
    const times = parseInt(req.query.times) || 1;
    //console.log(min, max, times);

    if (min == null || min < 0) { //this allow min to be 0
      return undefined;
    }
    if (!max || min > max || max > 10000000000) {
      return undefined;
    }
    if (times > 10000) {
      return undefined;
    }

    this.clean();
    if (times === 1) {
      this.response.quest = `Random Integer between ${min}-${max}`;
    } else {
      this.response.quest = `${times} Random Integers between ${min}-${max}`;
    }
    for (let i = 0; i < times; i++) {
      this.response.data.push(aux.randomInt(min, max));
    }
    return this.response;
  },
  list: function (req) {
    const elements = parseInt(req.query.len);
    //console.log( elements);

    if (!elements || elements < 2 || elements > 10000) {
      return undefined;
    }

    this.clean();
    this.response.quest = `Randomized order list with ${elements} elements`;
    let element = 0;
    while (element < elements) {
      const number = aux.randomInt(1, elements);
      if (!this.response.data.includes(number)) {
        this.response.data.push(number);
        element++;
      }
    }
    return this.response;
  }
};

File: random/random.go
/* */

package random

import (
	"fmt"
	"log"
	"net/http"
	"strconv"
	"strings"

	u "github.com/jolav/codetabs/_utils"
)

var surnameList []string

const (
	SURNAMES_FILE_PATH = "_data/random/surnames.txt"
)

func init() {
	fmt.Println("init Random")
	var err error
	surnameList, err = u.ReadFileLineByLine(SURNAMES_FILE_PATH)
	if err != nil {
		log.Print("INIT ERROR => Surnames files not found")
		surnameList = []string{}
	}
	fmt.Sprintln(surnameList)
}

type randomOLD struct {
	Quest  string `json:"quest"`
	Result []int  `json:"data"`
}

type random struct {
	Quest  string `json:"quest"`
	Result []int  `json:"data,omitempty"`
	Name   string `json:"name,omitempty"`
}

func Router(w http.ResponseWriter, r *http.Request) {
	params := strings.Split(strings.ToLower(r.URL.Path), "/")
	path := params[1:]
	if path[len(path)-1] == "" { // remove last empty slot after /
		path = path[:len(path)-1]
	}
	//log.Printf("Going ....%s %s %d", path, r.Method, len(path))
	if len(path) < 2 || path[0] != "v1" {
		u.BadRequest(w, r)
		return
	}
	if len(path) != 3 {
		u.BadRequest(w, r)
		return
	}

	if r.Method == "POST" {
		u.BadRequest(w, r)
		return
	}

	rd := newRandom()
	randomType := path[2]
	switch randomType {
	case "integer":
		rd.randomInteger(w, r)
	case "list":
		rd.randomList(w, r)
	case "name":
		rd.randomName(w, r)
	default:
		u.BadRequest(w, r)
		return
	}
}

func (rd random) randomInteger(w http.ResponseWriter, r *http.Request) {
	times := 1

	r.ParseForm()
	min, err := strconv.Atoi(r.Form.Get("min"))
	if err != nil {
		u.BadRequest(w, r)
		return
	}
	max, err := strconv.Atoi(r.Form.Get("max"))
	if err != nil || min > max || min < 0 || max > 10000000000 {
		u.BadRequest(w, r)
		return
	}
	if r.Form.Get("times") != "" {
		times, err = strconv.Atoi(r.Form.Get("times"))
		if err != nil || times > 10000 {
			u.BadRequest(w, r)
			return
		}
	}
	if times == 1 {
		rd.Quest = fmt.Sprintf("Random Integer between %d-%d", min, max)
	} else {
		rd.Quest = fmt.Sprintf("%d Random Integers between %d-%d", times, min, max)
	}

	for time := 1; time <= times; time++ {
		rd.Result = append(rd.Result, u.GetRandomInt(min, max))
	}

	u.SendJSONToClient(w, rd, 200)
}

func (rd random) randomList(w http.ResponseWriter, r *http.Request) {
	r.ParseForm()
	elements, err := strconv.Atoi(r.Form.Get("len"))
	if err != nil || elements < 2 || elements > 10000 {
		u.BadRequest(w, r)
		return
	}

	rd.Quest = fmt.Sprintf("Randomized order list with %d elements", elements)
	element := 0
	for element < elements {
		number := u.GetRandomInt(1, elements)
		if !u.SliceContainsInteger(number, rd.Result) {
			rd.Result = append(rd.Result, number)
			element++
		}
	}
	u.SendJSONToClient(w, rd, 200)
}

func (rd random) randomName(w http.ResponseWriter, r *http.Request) {
	pos := u.GetRandomInt(0, len(surnameList)-1)
	rd.Quest = "Random Name"
	rd.Name = surnameList[pos]
	u.SendJSONToClient(w, rd, 200)
}

func newRandom() random {
	rd := random{
		Quest:  "",
		Result: []int{},
		Name:   "",
	}
	return rd
}
